include/flang/semantics/tools.h

```diff
@@ -261,6 +261,55 @@ template<typename T> std::optional<std::int64_t> GetIntValue(const T &x) {
   }
 }
 
+template<typename T> std::optional<bool> GetBoolValue(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    return evaluate::ToBool(*expr);
+  } else {
+    return std::nullopt;
+  }
+}
+
+template<typename T> std::optional<std::string> GetStringValue(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    if (const auto *charExpr{
+            evaluate::UnwrapExpr<evaluate::Expr<evaluate::SomeCharacter>>(
+                *expr)}) {
+      return ToString(
+          std::get<evaluate::Expr<evaluate::Type<TypeCategory::Character, 1>>>(
+              (*charExpr).u));
+    }
+  }
+  return std::nullopt;
+}
+
+template<typename T>
+std::optional<std::u16string> GetU16StringValue(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    if (const auto *charExpr{
+            evaluate::UnwrapExpr<evaluate::Expr<evaluate::SomeCharacter>>(
+                *expr)}) {
+      return ToString(
+          std::get<evaluate::Expr<evaluate::Type<TypeCategory::Character, 2>>>(
+              (*charExpr).u));
+    }
+  }
+  return std::nullopt;
+}
+
+template<typename T>
+std::optional<std::u32string> GetU32StringValue(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    if (const auto *charExpr{
+            evaluate::UnwrapExpr<evaluate::Expr<evaluate::SomeCharacter>>(
+                *expr)}) {
+      return ToString(
+          std::get<evaluate::Expr<evaluate::Type<TypeCategory::Character, 4>>>(
+              (*charExpr).u));
+    }
+  }
+  return std::nullopt;
+}
```

Comment:
Can these 3 be a single function?

Reply:
Really I tried my best to combine these 3 functions into one. But I couldn't. This was the reason, I had earlier commented asking for any help in extracting these 3 values without conversion. Do you have any suggestions how I can make this a single function.?

Reply:
I will come back to you on this.

Reply:
You can write the functions the same way as follows:
```c++
template<typename S = std::string, typename T> 
std::optional<S> GetStringValue(const T &x) {
  if (const auto *expr{GetExpr(x)}) {
    if (const auto *charExpr{
            evaluate::UnwrapExpr<evaluate::Expr<evaluate::SomeCharacter>>(
                *expr)}) {
      return ToString(
          std::get<evaluate::Expr<evaluate::Type<TypeCategory::Character, sizeof(typename S::value_type)>>>(
              (*charExpr).u));
    }
  }
  return std::nullopt;
}
```
and then change the calls to the non-std::string ones from 
`GetU16StringValue` -> `GetStringValue<std::u16string>`
and 
`GetU32StringValue` -> `GetStringValue<std::u32string>`

Reply:
Thanks @DavidTruby.

Reply:
Thank you @DavidTruby 
I will use your implementation.